conans/client/cmd/uploader.py

```diff
@@ -83,28 +84,48 @@ def _upload(self, conan_file, conan_ref, force, packages_ids, retry, retry_wait,
         if not force:
             self._check_recipe_date(conan_ref, upload_remote)
 
-        self._user_io.out.info("Uploading %s to remote '%s'"
-                               % (str(conan_ref), upload_remote.name))
-        self._upload_recipe(conan_ref, retry, retry_wait, skip_upload, no_overwrite, upload_remote)
+        self._user_io.out.info("Uploading %s to remote '%s'" % (str(conan_ref), upload_remote.name))
+        new_ref = self._upload_recipe(conan_ref, retry, retry_wait, skip_upload, no_overwrite,
+                                      upload_remote)
 
-        recorder.add_recipe(str(conan_ref), upload_remote.name, upload_remote.url)
+        if not skip_upload:
+            recorder.add_recipe(new_ref.full_repr(), upload_remote.name, upload_remote.url)
 
         if packages_ids:
+            # Filter packages that don't match the recipe revision
+            if new_ref.revision:
+                recipe_package_ids = []
+                for package_id in packages_ids:
+                    pref = PackageReference(new_ref, package_id)
+                    package_folder = self._client_cache.package(pref, conan_file.short_paths)
+                    rec_rev = ConanInfo.load_from_package(package_folder).recipe_revision
+                    if new_ref.revision != rec_rev:
+                        self._user_io.out.warn("Skipping package '%s', it doesn't belong to "
+                                               "the current recipe revision" % package_id)
+                    else:
+                        recipe_package_ids.append(package_id)
+                packages_ids = recipe_package_ids
+
             # Can't use build_policy_always here because it's not loaded (only load_class)
             if conan_file.build_policy == "always":
                 raise ConanException("Conanfile has build_policy='always', "
                                      "no packages can be uploaded")
             total = len(packages_ids)
             for index, package_id in enumerate(packages_ids):
-                ret_upload_package = self._upload_package(PackageReference(conan_ref, package_id),
+                ret_upload_package = self._upload_package(PackageReference(new_ref, package_id),
                                                           index + 1, total, retry, retry_wait,
                                                           skip_upload, integrity_check,
                                                           no_overwrite, upload_remote)
                 if ret_upload_package:
-                    recorder.add_package(str(conan_ref), package_id)
+                    recorder.add_package(new_ref.full_repr(), package_id)
+
+        if skip_upload:
+            return
 
-        if not defined_remote and not skip_upload:
-            self._registry.set_ref(conan_ref, upload_remote.name)
+        if not defined_remote:
+            self._registry.set_ref(new_ref, upload_remote.name)
+        elif defined_remote == upload_remote:
```

Comment:
This is weird, why not an OR? And why if it is already defined has to be ``set_ref``?

Reply:
yes, it is an OR. about the set_ref, in the previous merged refactor the set_ref now take care of the update_ref if needed.